### PR TITLE
Add `$cluster` client property to annotation objects

### PR DESCRIPTION
**Pull request by lyzadanger**
This property is set when an annotation object is initialized in:

* annotator: In `guest`, when a user creates a new annotation or highlight via the adder controls
* sidebar: In the `annotations` store module when annotation objects are initialized before being added to the store

This property is communicated between the sidebar and the annotator when exchanging annotation data. In the annotator, the value of this property is used to set an additional CSS class on drawn anchor highlights (`<hypothesis-highlight>` `span`s).

The presence of this CSS class will allow subsequent differentiated styling for highlight clusters.

Impact: Every `hypothesis-highlight` span will have an additional CSS class corresponding to its `$cluster`. However, there is no visible change as we are not applying any styles for those classes (yet).

Part of #4916

### CODE REVIEW

File: src/annotator/guest.js [hunk 1]
```diff
@@ -543,7 +545,13 @@ export class Guest {
       }
 
       const highlights = /** @type {AnnotationHighlight[]} */ (
-        highlightRange(range)
+        highlightRange(
+          range,
+          /* cssClass */ classnames(
```
Comment (robertknight) [inline src/annotator/guest.js:550]: I think the purpose of this parameter can be inferred from how it is generated.

File: src/annotator/test/guest-test.js [hunk 2]
```diff
@@ -1125,6 +1137,22 @@ describe('Guest', () => {
       ]);
     });
 
+    it('provides CSS classes for anchor highlight elements', () => {
+      const guest = createGuest();
+      const annotation = {
+        $cluster: 'user-annotations',
+        target: [{ selector: [{ type: 'TextQuoteSelector', exact: 'hello' }] }],
+      };
+      fakeIntegration.anchor.returns(Promise.resolve(range));
```
Comment (robertknight) [inline src/annotator/test/guest-test.js:1146]:
```suggestion
      fakeIntegration.anchor.resolves(range);
```

File: src/annotator/test/guest-test.js [hunk 3]
```diff
@@ -1125,6 +1137,22 @@ describe('Guest', () => {
...
+
+      return guest.anchor(annotation).then(() => {
```
Comment (robertknight) [inline src/annotator/test/guest-test.js:1148]: This could use `await` to avoid the promise chain.

File: src/types/shared.ts [hunk 4]
```diff
@@ -1,7 +1,18 @@
...
+export type HighlightCluster =
+  | 'other-content' // default cluster: content not belonging to the current user
+  | 'user-annotations' // An annotation belonging to the current user
+  | 'user-highlights'; // A highlight (highlights are private; they always belong to the current user)
+
 /**
  * Annotation properties not present on API objects, but added by the client
  */
 export type ClientAnnotationData = {
+  $cluster?: HighlightCluster;
```
Comment (robertknight) [inline src/types/shared.ts:15]: The use of a `$cluster` property on annotations works for the current highlight clusters, because an annotation's cluster never changes once the `Annotation` object has been loaded/created. If in future we add new clustering methods which result in the annotation changing clusters dynamically then we'll either need to modify all the annotation objects (which could have performance overhead from invalidating various cached data, or just cache invalidation issues if we forget to update `$cluster` in some context), or we might need some system of having layers of cluster assignments, with a mechanism to resolve which one is active.

**Review by robertknight — APPROVED** [review]
LGTM. I did note some potential hazards / issues with using a `$cluster` property on annotations as we introduce new clustering modes in future which can change dynamically. Have you thought about how you might handle that?

File: src/sidebar/services/frame-sync.ts [hunk 5]
```diff
@@ -37,8 +37,14 @@ type DocumentInfo = {
  * JavaScript, it includes only the information needed to uniquely identify it
  * within the current session and anchor it in the document.
  */
-export function formatAnnot({ $tag, target, uri }: Annotation): AnnotationData {
+export function formatAnnot({
+  $cluster,
+  $tag,
+  target,
+  uri,
+}: Annotation): AnnotationData {
```
Comment (robertknight) [inline src/sidebar/services/frame-sync.ts:45]: It is currently the case that all the data we sent to the annotator for an annotation comes from the sidebar's internal `Annotation` object, but we're not obliged to keep things that way in future. We could for example look up additional information from other parts of the store and combine the results to generate the object sent to the annotator.

Comment (lyzadanger) [reply src/sidebar/services/frame-sync.ts:45]: Noted.

I think what I'd like to aim for is:

* There is a single annotation-object initialization point on both sides of the app (`annotator` and `sidebar`). This initialization logic may not have all of the answers — e.g. maybe we could initialize client-side properties to some default value — but I think there's value in having those properties at least initially set on the object in one place.
* We spend some time in the near future cleaning up some typing technical debt around `Annotation` — I'd like to more clearly separate API annotation typing from "app-wide" merged client-side properties versus what a constructed annotation object looks like in the `annotator` vs. what an annotation object looks like in the `sidebar` (these things are all different in reality). We might opt to think about changing the naming or structure of client-side properties, as you have suggested, at that time. That will help clarify this situation, I believe.

One future use case we've briefly discussed is having the sidebar's annotation cards have visual continuity with the annotation's cluster, e.g. making the highlight/quote area echo the color of the associated cluster. For that, the sidebar will need `$cluster` information, so I would like to leave the `$cluster` logic in the `annotations` store module _for now_ but will keep this all in mind...

File: src/types/shared.ts (hunk 4 above)
Comment (lyzadanger) [reply src/types/shared.ts:15]: Let's talk about this with voices in our next project check-in, but yes, this is important and something I've been keeping in mind as I go here. Very broadly, I could foresee:

* Static clusters known to the app (like this feature). Annotations either don’t change which cluster they’re in once loaded/created, or changes to an individual annotation’s cluster would happen if the annotation were modified. Changes to a single annotation’s cluster in the sidebar could be communicated to the annotator[^1] but should allow an existing highlight to be updated (i.e. CSS class changed or something), not redrawn. And certainly not require re-anchoring.
* Dynamic clustering. We’ve talked about a possible “emphasis mode”, for example, in which, say, search results or other sets of annotations could be visually intensified while other highlights would visually recede. I believe we could handle this similarly to how we currently handle focusing an annotation: the sidebar could pass the annotator a set of tags; the annotator could toggle a CSS class for highlight elements corresponding to those tags. Modally-applied styling would disable other cluster styling and make any highlight elements without the emphasis class visually recede. CSS class toggling — which we’re already doing for focused highlights — not re-drawing anchors.

tl;dr: Unless I do something really wrong, we should never need to modify multiple annotation objects or re-draw anchor highlights, and an annotation's `$cluster` value should only change on create/load or modify/save. We do need to support the ability for an annotation's `$cluster` to change when the annotation itself is modified.

[^1]: In thinking through how to explain this, I realized there is a small bug I'll need to kick out of the current implementation of this feature: in the slightly uncommon but not rare situation that a user edits a highlight and adds body text, it becomes an annotation. We need a mechanism to a) update an annotation's `$cluster` appropriately on modify/save (probably save) in the sidebar and b) communicate that change to the annotator. I'm adding that to my queue.
